source/Ship.cpp

```diff
@@ -3907,20 +3915,97 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
 // cues and try to stay with it when it lands or goes into hyperspace.
 void Ship::AddEscort(Ship &ship)
 {
-	escorts.push_back(ship.shared_from_this());
+	auto sp = ship.shared_from_this();
+	escorts.list.push_back(std::move(sp));
+	TuneForEscort(ship);
 }
 
 
 
 void Ship::RemoveEscort(const Ship &ship)
 {
-	auto it = escorts.begin();
-	for( ; it != escorts.end(); ++it)
-		if(it->lock().get() == &ship)
+	// When removing the slowest escort, we need to determine the new "speed limit."
+	shared_ptr<Ship> slowest = escorts.slowest.lock();
+	const bool findSlowest = (!slowest || (&ship == slowest.get()));
+	if(findSlowest)
+	{
+		escorts.cruiseVelocity = -1.;
+		escorts.slowest.reset();
+	}
+
+	auto it = escorts.list.begin();
+	while(it != escorts.list.end())
+	{
+		auto escort = it->lock();
+		if(escort.get() == &ship)
 		{
-			escorts.erase(it);
-			return;
+			it = escorts.list.erase(it);
+			// If we are not removing the slowest escort, then we don't
+			// need to finish the loop to tune all other escorts.
+			if(!findSlowest)
+				return;
+		}
+		else
+		{
+			if(escort && findSlowest)
+				TuneForEscort(*escort);
+			++it;
 		}
+	}
+}
+
+
+
+// (Re)Register escorts, for example because some escort got destroyed.
+void Ship::TuneForEscorts()
+{
+	// If the cached slowest escort is still ours, we don't need to re-find one.
+	shared_ptr<Ship> slowest = escorts.slowest.lock();
+	if(slowest && !slowest->IsDestroyed() && government == slowest->GetGovernment())
+		return;
+
+	escorts.cruiseVelocity = -1.;
+	escorts.slowest.reset();
+
+	for(const auto &it: escorts.list)
+	{
+		auto escort = it.lock();
+		if(escort)
+			TuneForEscort(*escort);
+	}
+}
+
+
+
+// Cache the maximum speed that the parent should stay below to keep
+// all escorts together.
+void Ship::TuneForEscort(const Ship &ship)
+{
+	// We don't cache the speeds of carried ships, since they are often
+	// docked and the carrier waits for them during docking already.
+	if(!ship.CanBeCarried() && !ship.IsDestroyed() && government == ship.GetGovernment())
+	{
+		double eV = ship.MaxVelocity() * 0.9;
+		// We don't cache the speeds of escorts that have 0 velocity to avoid
+		// having the parent get stuck when one of the escorts doesn't have
+		// regular thrust.
+		if(eV > 0. && (escorts.cruiseVelocity <= 0. || (eV < escorts.cruiseVelocity)))
+		{
+			escorts.cruiseVelocity = eV;
+			auto sp = make_shared<Ship>(ship);
+			escorts.slowest = sp;
```

Comment:
Invoking the Ship copy constructor is _definitely not_ what you want.
That will lead to a totally new shared_ptr, rather than incrementing the refcount:

```suggestion
			escorts.slowest = ship.shared_from_this();
```